slice/skip/skip.go

```diff
@@ -171,9 +260,44 @@ func (sl *SkipList) Insert(entries ...Entry) Entries {
 	return overwritten
 }
 
+func (sl *SkipList) insertAtPosition(position uint64, entry Entry) {
+	if position > sl.num {
+		position = sl.num
+	}
+	sl.cache.reset()
+	sl.posCache.reset()
+	n, pos := sl.searchByPosition(position, sl.cache, sl.posCache)
+	insertNode(sl, n, entry, pos, sl.cache, sl.posCache, true)
+}
+
+// InsertAtPosition will insert the provided entry and the provided position.
+// If position is greater than the length of the skiplist, the entry
+// is appended.  This method bypasses order checks and checks for
+// duplicates so use with caution.
+func (sl *SkipList) InsertAtPosition(position uint64, entry Entry) {
+	sl.insertAtPosition(position, entry)
+}
```

Comment:
It has to be cleared on every call because the cache is specific to an individual insertion operation (the cache is being used to track path information basically).  A faster option would probably be to just allocate on every insertion but I was concerned about memory consumption especially when skiplists already can be somewhat memory intensive (every node has two lists attached to it now).  The best optimization is probably a pool-based cache, target for further performance enhancements.